From 71ad7079afe00ddd7cfda291b9909de581e4a358 Mon Sep 17 00:00:00 2001
From: Stuart Morgan <smorgan@mythtv.org>
Date: Sat, 22 Feb 2014 08:59:54 +0000
Subject: [PATCH] HTTP Server: Meet HTTP/1.1 requirement that a 400 status
 code be given if the client failed to send a Host header.

---
 mythtv/libs/libmythupnp/httprequest.cpp |   28 ++++++++++++++++++++++++----
 mythtv/libs/libmythupnp/httpserver.cpp  |    3 ++-
 2 files changed, 26 insertions(+), 5 deletions(-)

diff --git a/mythtv/libs/libmythupnp/httprequest.cpp b/mythtv/libs/libmythupnp/httprequest.cpp
index bb80e84..5d11928 100644
--- a/mythtv/libs/libmythupnp/httprequest.cpp
+++ b/mythtv/libs/libmythupnp/httprequest.cpp
@@ -111,12 +111,21 @@ static MIMETypes g_MIMETypes[] =
     { "ts"  , "video/mp2t"                 },
 };
 
-static const char *Static401Error = 
-    "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\""
-        "\"http://www.w3.org/TR/1999/REC-html401-19991224/loose.dtd\">"
+static const char *Static400Error =
+    "<!DOCTYPE html>"
     "<HTML>"
       "<HEAD>"
-        "<TITLE>Error</TITLE>"
+        "<TITLE>Error 400</TITLE>"
+        "<META HTTP-EQUIV=\"Content-Type\" CONTENT=\"text/html; charset=ISO-8859-1\">"
+      "</HEAD>"
+      "<BODY><H1>400 Bad Request.</H1></BODY>"
+    "</HTML>";
+
+static const char *Static401Error =
+    "<!DOCTYPE html>"
+    "<HTML>"
+      "<HEAD>"
+        "<TITLE>Error 401</TITLE>"
         "<META HTTP-EQUIV=\"Content-Type\" CONTENT=\"text/html; charset=ISO-8859-1\">"
       "</HEAD>"
       "<BODY><H1>401 Unauthorized.</H1></BODY>"
@@ -1141,6 +1150,17 @@ bool HTTPRequest::ParseRequest()
             return false;
         }
 
+        // HTTP/1.1 requires that the Host header be present, even if empty
+        if ((m_nMinor == 1) && !m_mapHeaders.contains("host"))
+        {
+            m_eResponseType   = ResponseTypeHTML;
+            m_nResponseStatus = 400;
+
+            m_response.write( Static400Error );
+
+            return true;
+        }
+
         m_bProtected = false;
 
         if (IsUrlProtected( m_sBaseUrl ))
diff --git a/mythtv/libs/libmythupnp/httpserver.cpp b/mythtv/libs/libmythupnp/httpserver.cpp
index 0d77af4..4fedc2a 100644
--- a/mythtv/libs/libmythupnp/httpserver.cpp
+++ b/mythtv/libs/libmythupnp/httpserver.cpp
@@ -308,7 +308,8 @@ void HttpWorker::run(void)
                         // delegate processing to HttpServerExtensions.
                         // ------------------------------------------------------
 
-                        if (pRequest->m_nResponseStatus != 401)
+                        if ((pRequest->m_nResponseStatus != 400) ||
+                            (pRequest->m_nResponseStatus != 401))
                             m_httpServer.DelegateRequest(pRequest);
                     }
                     else
-- 
1.7.10.2

